Use safepoint to deliver SIGINT

yuyichao opened the pull request:
This implements @vtjnash 's [idea](https://github.com/JuliaLang/julia/issues/14675#issuecomment-171810732) of delivering sigint in a safer way, without having to add sigatomic to everywhere. This also solves a few performance issues along the way. A brief summary of the differences,
- Make GC safepoint thread local
  
  And create 3 signal pages. See the comment in `safepoint.c` for detail.
  This have a very minor performance hit for accessing safepoint. However, I think we can avoid the safepoint and GC transition around gcframe setup and gcframe pop on x86 so this shouldn't be a big issue. (On ARM, the atomic release store is more expensive than two GC safepoint so we may use a normal store with two safepoint instead, the performance hit of making the safepoint thread local is still very minor in this case though (~5-10% for empty GC frame push/pop, cheaper than the tls getter call....)).
- Use the GC safepoint mechanism (i.e. SegFault) to deliver InterruptException at known points
  
  This automatically makes `ccall` of external library `sigatomic`. (Fix #1468, Fix #2622). In order to avoid waiting too long before the exception is delivered when running unmanaged code, this also implement waking up libuv, abort syscall and abort libsupport io (this use a hack but can be generalized if needed) when sigint arrives. This also implement force throwing of exception if SIGINT arrives too frequently so that one can use `Ctrl-C` to force abort some dead loop. A warning will be printed in such case since it bypass the safe path and even sigatomic.
  
  The most important consequence IMHO is that pressing `Ctrl-C` during sysimg compilation should no longer segfault =)
  
  This should also make it easier to implement #14675.
- Make parser and type inference calls sigatomic
  
  sigatomic is much cheaper (see below) now so it is now used to protect important runtime code to not be interrupted by `Ctrl-C`.
- Clean up safepoint and sigatomic, optimize try-catch and `sigatomic`
  - Split out `safepoint.c`, make `defer_signal` thread local and avoid the expensive atomic ops
  - Make `defer_signal` task local and automatically restore on expection. This eliminate the try-catch needed before in `disable_sigint`.
    
    This is technically breaking but `sigatomic_begin` and `sigatomic_end` aren't exported.
  - Inline `jl_sigatomic_begin` and `jl_sigatomic_end` in codegen.
  
  Benchmarking try-catch and `disable_sigint`
  
  ``` jl
  julia> @inline f1() = try end
  f1 (generic function with 1 method)
  
  julia> @inline f2() = disable_sigint(()->nothing)
  f2 (generic function with 1 method)
  
  julia> g1(n) = for i in 1:n
             f1()
         end
  g1 (generic function with 1 method)
  
  julia> g2(n) = for i in 1:n
             f2()
         end
  g2 (generic function with 1 method)
  
  julia> function k(n)
             @time g1(n)
             @time g2(n)
         end
  k (generic function with 1 method)
  ```
  
  Timing before
  
  ``` jl
  julia> k(10_000_000)
    0.264041 seconds
    0.401437 seconds
  ```
  
  Timing after
  
  ``` jl
  julia> k(10_000_000)
    0.122592 seconds                                                                
    0.012506 seconds
  ```

Close https://github.com/JuliaLang/julia/pull/12333
Close #12309

yuyichao committed:
Enable GC safepoint for non-threading build

yuyichao committed:
Split out safepoint.c

Add note and function prototypes about using safepoint to deliver SIGINT.

yuyichao committed:
Make safepoint thread local.

Preparing for using the safepoint to deliver SIGINT, which requires distinguish
between master and worker threads.

yuyichao committed:
Use safepoint to deliver SIGINT

* Remove unnecessary sigatomic
* Make flisp calls sigatomic
* Make type inference calls sigatomic
* Refactor interthread communication through signal
* Make sure `sleep` is aborted on `SIGINT` on Linux to deliver the exception faster
* Implement force signal throwing when `SIGINT` arrives too frequently
* Hack to abort io syscall on `SIGINT`

Fix #1468; Fix #2622; Towards #14675

yuyichao committed:
Restore delay_sigint on exception and task switch

Also add tests.

yuyichao committed:
Inline `jl_sigatomic_begin` and `jl_sigatomic_end`

yuyichao committed:
Add atomic exchange functions

yuyichao committed:
Wake up libuv main loop on SIGINT

vtjnash commented on src/julia.h:
```diff
@@ -1494,7 +1494,13 @@ STATIC_INLINE void jl_eh_restore_state(jl_handler_t *eh)
+    // that throws a SIGINT.
+    jl_gc_state_save_and_set(eh->gc_state);
+    if (eh->defer_signal == 0) {
+        jl_sigint_safepoint();
```
i'm fairly certain we don't actually want `catch` to be a safepoint (now that it's avoidable)

yuyichao replied:
The sigint safepoint is kind of avoidable (by reenable the gc safepoint) I don't think it's valid to avoid the GC safepoint in the gc state restore though.

vtjnash replied:
that's fine. i just don't think we want to throw an InterruptException before the catch block has a chance to install a try/catch of its own.

yuyichao replied:
We can't really provide the grantee that an async signal will not be thrown between a `catch` and a `try` though, unless we garentee it can only be compiled and not interpreted. The GC safepoint can also throw the exception even when there's not a sigint safepoint.

tkelman commented on src/julia_internal.h:
```diff
@@ -264,10 +265,63 @@ void jl_set_base_ctx(char *__stk);
+
+// Whether the GC is running
+extern char *jl_safepoint_pages;
+static inline int jl_addr_is_safepoint(uintptr_t addr)
```
`STATIC_INLINE`

yuyichao replied:
What configuration doesn't support `static inline`? It seems that MSVC should be fine with it in c++ mode?

tkelman replied:
we won't always need to compile in c++ mode

yuyichao replied:
I fixed this one. However, there are a lot of threading utils that relies on c++ when compiled with MSVC. Alternative would be using C11 generic macro, which I think is even less likely going to be supported by MSVC....

vtjnash commented on src/signals-unix.c:
```diff
@@ -131,72 +140,80 @@ static void allocate_segv_handler(void)
-        wait_barrier();
+        request = jl_atomic_exchange(&ptls->signal_request, 0);
+        assert(request == 1);
+        if (request)
```
assert & conditional seems like an odd pair?

yuyichao replied:
Fixed. I think the two was added at different point and I didn't remove them in rebase/squash...

vtjnash commented on src/signals-unix.c:
```diff
@@ -131,72 +140,80 @@ static void allocate_segv_handler(void)
+                jl_safe_printf("WARNING: Force throwing a SIGINT\n");
+            // Force a throw
+            jl_clear_force_sigint();
+            jl_throw(jl_interrupt_exception);
```
maybe add a TODO to replace this with `jl_throw_in_ctx` (after writing said function)

yuyichao replied:
TODO with `Jameson` attached added =)

vtjnash commented on src/task.c:
```diff
@@ -318,6 +318,8 @@ static void ctx_switch(jl_task_t *t, jl_jmp_buf *where)
             arraylist_new(locks, 0);
         }
 #endif
```
i think FIXME comment at the top of this function can be removed now?

yuyichao replied:
Done.

stevengj commented:
Suppose I have a `ccall` with a Julia callback function, e.g. `ccall(:my_numeric_integration, Cdouble, (Ptr{Void},), cfunction(my_julia_integrand, ...))`.    If I send a SIGINT while this is executing, does this mean that it will not interrupt the C code, but an `InterruptException` will still occur during the `my_julia_integrand` execution?

Also, if you have multiple `Task`s running, is there any way to indicate which `Task` you want to catch the exception?

yuyichao commented:
A callback can still throw exception so sigatomic_begin/end are still useful. I made it fast because of this...

The sigint is always delivered to current task on the master thread if it doesn't block sigint (I realized that I might have missed a pending sigint check there, similar to the one in en_restore_state). However, you can block the sigint on the task you don't want it to be delivered by calling sigatomic_begin, is this good enough?

stevengj commented:
I'm still confused by the callback situation in this PR.  If I *want* `InterruptExceptions` to be thrown during my callback (e.g. I have a `try-catch` block that handles them correctly), do I need to call `sigatomic_end/begin` in my callback?

yuyichao commented:
For julia callback from C, I think a much faster sigatomic is as far as what this approach can provide.

For the task issue, I'm wondering if we can make non-root tasks sigatomic by default? This way any frontend will need to explicitly reenable sigint on tasks that want to handle sigint.

yuyichao commented:
> (e.g. I have a try-catch block that handles them correctly), do I need to call sigatomic_end/begin in my callback?

A sigint can be delivered in the callback if the C caller wasn't called in sigatomic region. It is a little unreliably though since in principle a sigint can be delivered before the try block starts so I think to do this reliably you need to call the C function with sigatomic and reenable sigatomic in the callback after you started the try block.

stevengj commented:
I'm not sure if non-root tasks should be sigatomic by default.  But it would be nice to have an option to make the task sigatomic when it is created.  e.g. `@async_nosigint`

I'm not sure how calling `sigatomic_begin/end` will help in Tasks, right now.  The problem is that `sigatomic_end` is still called within the Task, and as soon as it is called any deferred SIGINT will be thrown as an exception in that Task.  What if I don't want that Task to get `InterruptExceptions` at all?

stevengj commented:
> you need to call the C function with sigatomic and reenable sigatomic in the callback after you started the try block.

Yes, that's what I'm doing now.

yuyichao commented:
You can call sigatomic_begin without calling end to permanently block sigint on a task. Doing that at initialization time is also possible since I accidentally made defer_signal a property of the task (instead of just save it on the stack on task switch....)

vtjnash commented:
What is the sigatomic state of a new Task? e.g.
```julia
sigatomic_begin()
@async println("am I sigatomic?")
sigatomic_end()
yield()
```

stevengj commented:
@yuyichao, `sigatomic_begin()` will only block SIGINT on the current task?  I thought it increments `jl_defer_signal`, which is global to all tasks?

yuyichao commented:
sigatomic is now thread and task local. New task currently (on this branch) inherit the state from when it's started... since I haven't set fix start_task for this yet...

yuyichao committed:
Set GC state before calling debuginfo.cpp.

Depending on the caller these functions can be called from unmanaged threads
that cannot touch GC state or having GC safepoint.

yuyichao commented:
Travis linux32 failure is https://github.com/JuliaLang/julia/issues/15920

tkelman commented on src/gc.c:
```diff
@@ -354,95 +355,26 @@ NOINLINE static uintptr_t gc_get_stack_ptr(void)
-        while (!ptls->gc_state) {
+        // FIXME: The acquire load pairs with the release store
+        // in the signal handler of safepoint so we are sure that
+        // all the memory write on those threads are visible. However,
```
memory writes? memory written ... is?

yuyichao replied:
Maybe `store` is the right word?

tkelman commented on test/core.jl:
```diff
@@ -2142,13 +2142,24 @@ end
     ccall(:jl_exit_on_sigint, Void, (Cint,), 1)
 end
+let
+    # Exception frame automatically restore sigatomic counter.
```
restores?

yuyichao commented:
Spelling mistakes fixed. Threading CI passed [AppVeyor](https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.1224), [travis](https://travis-ci.org/JuliaLang/julia/builds/128043719). This should be ready for final review now.

stevengj commented:
It seems like the documentation of the `sigatomic` functions should be revisited.  (I'd also like to see documentation of how to catch ctrl-c in callback functions, disable ctrl-c in certain tasks, etcetera.)

JeffBezanson commented on src/signal-handling.c:
```diff
@@ -23,11 +23,35 @@ static const    uint64_t GIGA = 1000000000ULL;
+        accum_weight = 0;
+        return 0;
+    }
+    double new_weight = accum_weight * exp(-(dt / 1e9)) + 0.3;
```
Clearly written by a physicist :)

yuyichao commented:
I've also noticed one case where `Ctrl-C` can still segfault. It is when we throw the error without any handler, and the execution of `on_exit` overflows the signal handler stack..... (at least on linux where we throw directly from the signal handler frame). A easy fix would be setting up a top level try catch frame at initialization time. The alternative would be to return to a call to `jl_throw` from the signal handler instead. However, it seems that `setcontext` or `makecontext` using a signal context might be undefined....

JeffBezanson commented on src/ast.c:
```diff
@@ -267,6 +268,7 @@ static jl_ast_context_t *jl_ast_ctx_enter(void)
 
 static void jl_ast_ctx_leave(jl_ast_context_t *ctx)
 {
+    JL_SIGATOMIC_END();
```
This breaks `--lisp`; segfaults.

yuyichao replied:
Ah... ok.... will fix later today....

yuyichao replied:
And in case you need it now, I think making jl_sigint_safepoint noop should work around it for now

yuyichao replied:
Fixed in https://github.com/JuliaLang/julia/commit/417e49aae87222d96df6125b188686da39a9bc80

tkelman replied:
should be tested if breaking it matters

yuyichao replied:
`` run(pipeline(DevNull, `$(Base.julia_cmd()) --lisp`, DevNull)) `` ?

tkelman replied:
something like that. might need to be unix only, would need to run it through CI first

tkelman commented:
This PR broke the Windows buildbots: https://build.julialang.org/builders/package_win6.2-x64/builds/247/steps/make/logs/stdio

Similar to #11727, but now bootstrap freezes when output is redirected instead of just being unable to run tests. @vtjnash @yuyichao please advise, we'll have to revert this if no solution can be found.

yuyichao commented:
The only commit that touches libuv is the last one and reverting it seems to get rid of the problem for me locally (tested by piping it through `tee`) I'm checking now if what is the minimum change that triggers this now...

yuyichao commented:
Something is fishy here. The buildbot seems to have been printing `jl_uv_writecb` warnings forever during compilation http://buildbot.e.ip.saba.us:8010/builders/package_win6.2-x64/builds/245/steps/make/logs/stdio The freezing happens on current master after printing the first warning. However, either with or without the commit reverted, the build passes without printing any warnings if `inference.ji` exists before I run make. (i.e. if I run make again after it freezes, the build passes without any warning).

vtjnash commented:
If you follow that issue, you'll discover it's an upstream bug where libuv forgot to initialize the synchronization object. I think it then corrupts its internal state and tries to report an error that didn't happen by forgetting about a write req that did happen.

yuyichao commented:
> If you follow that issue

Which issue? I didn't see any mentioning of synchronization object in #11727?

vtjnash commented:
you have to follow it somewhat far. there's a more concise explanation at https://github.com/libuv/libuv/issues/629, in which I demonstrate how it would make nodejs to hang (for the same reason) in a one-liner.

yuyichao commented:
Ah, that one, I somehow think you are talking about `uv_async_t` since apparently simply initializing (not even using it) one is causing the trouble on windows......

tkelman commented:
I think this broke the arm build too: https://build.julialang.org/builders/package_tarballarm/builds/227/steps/make%20binary-dist/logs/stdio

yuyichao commented:
I believe it's caused by a llvm bug (probably similar to https://llvm.org/bugs/show_bug.cgi?id=27545) easy to workaround by removing an optimization, need to first check if it's fixed on llvm svn...
